packages/flutter_tools/lib/src/artifacts.dart

```diff
@@ -768,24 +768,22 @@ class CachedLocalEngineArtifacts implements LocalEngineArtifacts {
 
 
   @override
-  FileSystemEntity getHostArtifact(
-    HostArtifact artifact,
-  ) {
+  FileSystemEntity getHostArtifact(HostArtifact artifact) {
     switch (artifact) {
       case HostArtifact.engineDartSdkPath:
-        final String path = _fileSystem.path.join(_hostEngineOutPath, 'dart-sdk');
+        final String path = _getDartSdkPath();
         return _fileSystem.directory(path);
       case HostArtifact.engineDartBinary:
-        final String path = _fileSystem.path.join(_hostEngineOutPath, 'dart-sdk', 'bin', _hostArtifactToFileName(artifact, _platform));
+        final String path = _fileSystem.path.join(_getDartSdkPath(), 'bin', _hostArtifactToFileName(artifact, _platform));
```

Comment:
Do you think we should introduce a flag for this use case?

Reply:
Hold on. Maybe we should reverse the lookup order? So we check for a dart sdk inside the out folder. If we can't find it, we look in the prebuilts folder. What do you think?

Reply:
I think I might have been unclear. I think instead of this PR, the engine build should populate the directory that the tool is currently looking in.

Reply:
My previous comments were in response to the use case of a modified Dart SDK.

Reply:
> The output of the new --web build is platform-independent. We can build/zip it once then use the same on linux, mac and windows machines (in CI, for example).

Yeah, that makes sense. I guess the way it'd have to work is that if you say `--no-prebuilt-dart-sdk` for the `--web` build, you'd build the Dart SDK with the default host toolchain.

Maybe the way forward here is to file an issue for that, and for now reverse the lookup order as you suggest. I'd also like to know if @jonahwilliams sees any issues with that approach.

Reply:
Sounds good! I created an issue: https://github.com/flutter/flutter/issues/113772 and reversed the lookup order.

Reply:
I strongly believe we should be rethinking this idea in general, at least for the web build if not more broadly. The `--no-prebuilt-dart-sdk` flag really complicates our build process and I think we should in general be moving away from it. Supporting both paths makes things very complicated.

In the web case, it is weird to put native host artifacts such as the dart sdk inside of the wasm output directory. There are ways of invoking the host toolchain from the wasm build if we want, but I question whether this is really worth the effort to support this mode. More broadly, it feels wrong for the dart sdk to be an actual output of the flutter engine build, even in the native case.

Editing code in `third_party/dart` inside the engine repo doesn't really feel like a great workflow to me personally, I'm not sure how many people rely on this. If someone is making changes to the dart sdk, I expect that they should have the dart sdk repo itself already checked out and are able to build using the normal build process for the dart sdk. If anything we should probably add support for using an arbitrary built dart sdk instead of coupling that to the engine build.

I'm not sure what the best long-term plan here is, but I think the main reason we expect the dart sdk to be in the engine build directory is for historical reasons before we were mostly using the sdk prebuilts. This might be a good opportunity to think about how these dependencies work and seeing if we can improve this process.